### ISSUE TITLE
xmlrpc removed :(

### ISSUE THREAD

**Issue by insad** [webmin/webmin#1257]
Seems like xmlrpc code has been removed, completely breaking other code of us relying on this interface....

**Comment by insad** [comment]
Looks like some install problem with 1.951 - installing over 1.950 on (old Fedora) server seems to have crashed. Rolled back to 1.941 and everything working again.

**Comment by jcameron** [comment]
What went wrong with 1.951 exactly?

**Comment by insad** [comment]
Just installed version 1.952 without problems, but also xml-rpc api endpoints not working anymore.

E.g. a call to "useradmin::list_users" returns null, instead of an array of users like in version 1.941 (which I'm going to put back). 

No errors in miniserv.error / miniserv.log, just nothing returned anymore. So definitely there are api changes that have been made after 1.941 in at least the useradmin module (we use the useradmin and sendmail api's in our code, beside lock_file / unlock_file),

**Comment by iliarostovtsev** [comment]
>  but also xml-rpc api endpoints not working anymore.

If you open `miniserv.pl` file, find `$write_data_count += length($str);` around line 3168 and add before it:


~~~
select(undef, undef, undef, .0001) if ($write_data_count%10 == 0);
~~~

Restart Webmin afterwards by running `/etc/webmin/restart` does it change anything? If not, what if the value would be **.01** instead of **.0001**? Please try and report back!

**Comment by insad** [comment]
Tried both (including restart of the webmin service), but no luck....

**Comment by insad** [comment]
FYI I'm using SSL with a Letsencrypt certificate, and there are no errors in the log files, the xml-rpc request returns a 200 status.

**Comment by iliarostovtsev** [comment]
Do you have Perl module called XML::Parser installed?

**Comment by insad** [comment]
Yes it's installed (2.41-13.fc21) .... besides, downgrading to webmin version 1.941 everything works fine (using same parser I suppose), all later versions after 1.941 have the problem I mentioned. So it must be some problem introduced soon after launching version 1.941.

I'm attaching the - simplified - PHP code we use for calling the webmin API, maybe it can help you to debug the problem.

[server.zip](https://github.com/webmin/webmin/files/4875590/server.zip)

**Comment by iliarostovtsev** [comment]
Can you narrow down the issue? Install 1.952 and replace `miniserv.pl` file from 1.942 - does it work?

Try restarting Webmin running `pkill -9 miniserv`, and starting it `/etc/webmin/start.`.

**Comment by iliarostovtsev** [comment]
..You should only install 1.952 from the console, not UI.

**Comment by insad** [comment]
I think I found something.... comparing your miniserv.pl from the working and the new versions, I see you put in charset uft8 in many places. But the default xmlrpc_encode_request in PHP generates:

php > $x = xmlrpc_encode_request("useradmin::list_users", []);
php > echo $x;
<?xml version="1.0" encoding="iso-8859-1"?>
<methodCall>
<methodName>useradmin::list_users</methodName>
<params/>
</methodCall>
php >

So I think an encoding problem, will fix that first in my code, make a new test and let you know!

**Comment by iliarostovtsev** [comment]
Great. Thanks!

**Comment by iliarostovtsev** [comment]
`encoding="iso-8859-1"?` where is this line coming from? Can you get rid of it or try using `utf-8` instead?

**Comment by insad** [comment]
Changed the PHP code:

$request = xmlrpc_encode_request($function, $parameters, ["encoding" => "utf-8"]);

and it's sending the right encoding now, but even so not working :(

And yes, I install the 1.951 version from the console, after that I restart webmin (systemctl restart webmin)

**Comment by insad** [comment]
The request I'm sending now is

```
<?xml version="1.0" encoding="utf-8"?>
<methodCall>
<methodName>useradmin::list_users</methodName>
<params/>
</methodCall>
```

**Comment by insad** [comment]
Weird, testing with a direct curl command seems to work... maybe PHP doesn't like the miniserv response, will sort it out further....

**Comment by insad** [comment]
Definitely something going wrong inside PHP method xmlrpc_decode() - for some obscure reason doesn't like the response of the server... Will try to find out some other moment what exactly could be going on, no time for this now...

**Comment by iliarostovtsev** [comment]
We have switched to using `utf-8` completely, starting with Webmin 1.950. It might need additional tweaking on your app's side.
